darktable-org/darktable
# Bugfix: make sure correct HLR chroma corrections

See #13632 and #13646

In some situations the calculated chroma corrections are vastly wrong. The first idea was about precision errors because of using floats for summing up data and the counters, that could be ruled out.

The reported errors could be related (this is my current understanding) to:
1. improper data sampling because of the relaxed global memory policy, this could still be the case although unlikely because the errors keep happening after correcting data to be always in a gpu cacheline. (It might be we have to intruce a `lock-per-line`)
2. Other issues on AMD often pinpointed to a) float data not being initialized or b) math errors like div by zero. (Nvidia never had problems on this) The code has been reviewed for strictly avoiding this.
3. The for_each_channel usage was wrong in many parts of the code as we don't calc 4-channel pixels but we do a loop over a "color-range" of 0-2. This could have resulted to a div-by-zero.
4. The dark-level was a code left-over, it's safe to use 0.2 of clip in all cases. (This is the only maths change in this pr. The dark clips lead to problems in very few images if there were a lot of black-to-white transitions)
5. It is very good to avoid chroma corrections derived from a very small number of border pixels, we make sure now to have at least a "sensible" number.

Thanks to many people having tested #13646, could i ask you to check again @groutr @MStraeten @kofa73 @gi-man @sarunasb ?

To test this:
1. Use the "duck-at-the-pond" image from #13632 
2. Start dt either with `--disable-opencl` or not
3. use `-d pipe` and watch out for reports like `[opposed chroma cache GPU/CPU]`
4. Compare the values reported, they must be **"almost identical"** for CPU vs GPU, any difference must be concidered to be a bug.
5. If you made sure: rawprepare, temperature and clip in hlr are all set to default, and using opposed method your results should be
```
   13.2351 [opposed chroma cache GPU] red: 1.056767 (18406), green: -0.025674 (123257), blue: -0.394672 (16525) for opphash  11498776663581970802
    3.7372 [opposed chroma cache CPU] red: 1.056766 (18406), green: -0.025674 (123257), blue: -0.394673 (16525) for opphash  11498776663581970802
```

A sidenote, the algo has also changed slightly, the results might be slightly different from master but tested on this pr the reports must be identical for GPU and CPU
